api.c

```diff
@@ -0,0 +1,1055 @@
+        InArg,
+        InArgQuote,
+        OutOfArg,
+    };
+    enum State curr_state = OutOfArg;
+    char curr_quote = '\0';
+
+    while ((c = str[i]) != '\0') {
+        if (c == '\"' || c == '\'') { // Quote
+            switch (curr_state) {
+                case OutOfArg:
+                    free(curr_arg.value);
+                    init_str(&curr_arg, 20);
+                case InArg:
+                    curr_state = InArgQuote;
+                    curr_quote = c;
+                    break;
+                case InArgQuote:
+                    if (c == curr_quote)
+                        curr_state = InArg;
+                    else
+                        str_push_c(&curr_arg, c);
+                    break;
+            }
+        } else
+        if (c == ' ' || c == '\t') { // Whitespace
+            switch(curr_state) {
+                case InArg:
+                    push_back(&out_args, curr_arg);
+                    curr_state = OutOfArg;
+                    break;
+                case InArgQuote:
+                    str_push_c(&curr_arg, c);
+                    break;
+                case OutOfArg:
+                    break;
+            }
+        } else {
+            switch(curr_state) {
+                case InArg:
+                case InArgQuote:
+                    str_push_c(&curr_arg, c);
+                    break;
+                case OutOfArg:
+                    free(curr_arg.value);
+                    init_str(&curr_arg, 20);
+                    str_push_c(&curr_arg, c);
+                    curr_state = InArg;
+                    break;
+            }
+        };
+        i++;
+    }
+
+    if (curr_state == InArg) {
+        push_back(&out_args, curr_arg);
+    } else if (curr_state == InArgQuote) {
+        return 1;
+    }
+
+    size_t argc = out_args.size;
+    char **argv = str_destruct(out_args);
+
+    free(curr_arg.value);
+    free(out_args.items);
+
+    data->argc = argc;
+    data->argv = argv;
+
+    return 0;
+}
+
+void *listen_api(void* _) {
+
+    if (access(api_fifo_path, F_OK) == 0) {
+        if (remove(api_fifo_path) != 0) {
+            fprintf(stderr, "Could not remove file \"%s\": %s\n",
+                    api_fifo_path, strerror(errno));
+        }
+    }
+    if (mkfifo(api_fifo_path, S_IRWXU) != 0) {
+        fprintf(stderr, "Could not create FIFO file \"%s\": %s\n",
+                api_fifo_path, strerror(errno));
+    }
+
+    FILE *fifo;
+    char c;
+    int i;
+    int times = 0;
+    char *text = (char *) malloc(50);
+    arg_data data;
+
+    DEBUG("Created fifo at \"%s\"\n", api_fifo_path);
+
+    while (1) {
+        strcpy(text, "command ");
+        i = strlen(text);
+
+        fifo = fopen(api_fifo_path, "rb");
+        if (fifo == NULL) {
+            fprintf(stderr, "Could not open FIFO file \"%s\": %s\n",
+                    api_fifo_path, strerror(errno));
+            break;
+        }
+        while ((c = fgetc(fifo)) != EOF) {
```

Comment:
Maybe process the parameters at new line instead of EOF? So a program can continuously write to the fifo without closing and reopening it.

Reply:
I don't think so. If the user uses a different program to write on FIFO, instead of `echo`, that doesn't put any newline at the end, i3lock won't ever close the file and so won't handle those messages.
EOF seems good to me.

Reply:
Both EOF and newline then? It's mianly to support this kind of program:

```sh
exec > i3lock-fifo
echo --some-param
sleep 1
echo --more param
```

Reply:
Oh, sure, that program won't work as expected. Newlines are handled [here](https://github.com/JezerM/i3lock-color/blob/7995b902922f571ea7d4f6ff363e72b04027164a/api.c#L1028-L1033), but are replaced with spaces. I'll change it.